### PR TITLE
chore: enable `noImplicitAny` and `noImplicitThis`

### PR DESCRIPTION
**Pull request by DominicKramer**
Note that `tsconfig.json` still has `declaration` specified as `false` since that is required to have `allowJs` specified as `true`, which is required for the tests since they have not been converted to Typescript yet.

### CODE REVIEW

File: src/google-apis/auth-client.ts [hunk 1]
```diff
@@ -18,7 +18,8 @@
 import * as common from '@google-cloud/common';
 const pkg = require('../../../package.json');
 import * as is from 'is';
-var isFunction = is.fn;
+// TODO: Address the error where `is` does not have a `fn` property
```
Comment (jinwoo) [inline src/google-apis/auth-client.ts]: Isn't it better to throw an error when `is.fn` doesn't exist?
```typescript
if (!is.fn) {
  throw new Error('is.fn is not defined');
}
var isFuntion = is.fn;
```

Same for similar places in this PR.

File: src/google-apis/auth-client.ts [hunk 2]
```diff
@@ -115,7 +116,7 @@ export class RequestHandler extends common.Service {
 
     var that = this;
     if (tryAuthenticate) {
-      this.authClient.getToken(function(err) {
+      this.authClient.getToken(function(err: {message: string}) {
```
Comment (jinwoo) [inline src/google-apis/auth-client.ts]: `function(err: Error)` would work.

File: src/configuration.ts [hunk 3]
```diff
@@ -233,7 +233,7 @@ export class Configuration {
       version = env.GAE_MODULE_VERSION;
     }
 
-    this._serviceContext.service = isString(service) ? service : 'node';
+    this._serviceContext.service = (isString(service) ? service : 'node')!;
```
Comment (kjin) [inline src/configuration.ts]: nit: Maybe cast `isString` to `(obj: any) => obj is string` type?

File: src/google-apis/auth-client.ts [hunk 4]
```diff
@@ -18,7 +18,8 @@
 import * as common from '@google-cloud/common';
 const pkg = require('../../../package.json');
 import * as is from 'is';
-var isFunction = is.fn;
+// TODO: Address the error where `is` does not have a `fn` property
+var isFunction = (is as {} as {fn: Function}).fn;
```
Comment (kjin) [inline src/google-apis/auth-client.ts]: Use `is.function`: `is.fn === is.function`

File: src/google-apis/auth-client.ts (hunk 2 above)
Comment (kjin) [reply src/google-apis/auth-client.ts]: Would an arrow function work here?

File: src/index.ts [hunk 5]
```diff
@@ -87,85 +96,100 @@ import {createLogger} from './logger';
...
  */
-function Errors(initConfiguration: ConfigurationOptions): void {
-  if (!(this instanceof Errors)) {
-    return new Errors(initConfiguration);
-  }
+export class Errors {
+  private _logger: Logger;
+  private _config: Configuration;
+  private _client: AuthClient;
+  report: (err: {}, request?: Request, additionalMessage?: string|{}, callback?: Callback|{}|string) => ErrorMessage;
```
Comment (kjin) [inline src/index.ts]: Is it possible to define `report` multiple times? ex. https://github.com/grpc/grpc-node/blob/9f0e496ad839eaeb5e09c349e742e9c844de3e25/packages/grpc-js-core/src/client.ts#L119

File: system-test/error-reporting.js [hunk 6]
```diff
@@ -522,7 +522,8 @@ describe('error-reporting', function() {
       },
       extraConfig || {}
     );
-    errors = require('../src/index.js')(config);
+    const {Errors} = require('../src/index.js');
```
Comment (kjin) [inline system-test/error-reporting.js]: Not sure if you're transpiling this js file, but I think this syntax doesn't work in Node 4

File: src/google-apis/auth-client.ts (hunk 1 above)
Comment (DominicKramer) [reply src/google-apis/auth-client.ts]: I think that is a good idea, but for these PRs, as much as I can, I don't want to make behavioral changes.  I will address the TODOs in follow-up PRs.

File: src/google-apis/auth-client.ts (hunk 2 above)
Comment (DominicKramer) [reply src/google-apis/auth-client.ts]: I will introduce arrow functions in an upcoming PR when I run `gts fix`.

File: src/configuration.ts (hunk 3 above)
Comment (DominicKramer) [reply src/configuration.ts]: That would work too.  In the long run I want to remove the `is*` methods where possible.

File: system-test/error-reporting.js (hunk 6 above)
Comment (DominicKramer) [reply system-test/error-reporting.js]: Yes, good catch.  I'll update it.

File: src/index.ts (hunk 5 above)
Comment (DominicKramer) [reply src/index.ts]: Yes, since it is a change in the number of args it should work.

File: src/google-apis/auth-client.ts (hunk 4 above)
Comment (DominicKramer) [reply src/google-apis/auth-client.ts]: Awesome.  I'll update it.

Comment (DominicKramer) [reply src/google-apis/auth-client.ts]: I think I'll leave this for now using `is.fn`.  In a follow-up PR we can update the code to use `is.function`.

File: system-test/error-reporting.js (hunk 6 above)
Comment (DominicKramer) [reply system-test/error-reporting.js]: Fixed

File: src/index.ts (hunk 5 above)
Comment (DominicKramer) [reply src/index.ts]: To prevent merge conflicts for the PRs, I think it would be easier to update this in a follow-up PR.

**Comment by DominicKramer** [comment]
PTAL

File: package.json [hunk 7]
```diff
@@ -59,7 +59,9 @@
     "@google-cloud/nodejs-repo-tools": "^2.0.11",
     "@types/express": "^4.11.0",
     "@types/hapi": "^16.1.11",
+    "@types/is": "0.0.18",
     "@types/koa": "^2.0.43",
+    "@types/lodash.has": "^4.5.3",
```
Comment (kjin) [inline package.json]: Nit: Seems like this is unused?

Comment (DominicKramer) [reply package.json]: I looked into it, and its actually used in `src/populate-error-message.ts`.

**Pull request merged**